### PR TITLE
javascript/gigasecond: automated comments for approval and disapproval

### PR DESCRIPTION
**Pull request by SleeplessByte**

### CODE REVIEW

File: automated-comments/javascript/gigasecond/dont_use_get_seconds.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Use [`Date#getTime`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTime)
+to get the number of milliseconds the _entire_ `Date` represents, instead of
+getting the `seconds` component of the `Date`. In general, [Unix/UTC time](https://en.wikipedia.org/wiki/Unix_time),
+as provided by `Date#getTime`, is preferred when dealing with dates, as it is
+not affected by _interpretation_ or locale (such as timezones), and it allows
+for date manupilation without needing to rely on roll-over.
```
Comment (ErikSchierboom) [inline automated-comments/javascript/gigasecond/dont_use_get_seconds.md]:
```suggestion
for date manipulation without needing to rely on roll-over.
```

> roll-over

Will students know what "roll-over" is?

File: automated-comments/javascript/gigasecond/prefer_extracted_top_level_constant.md [hunk 2]
```diff
@@ -0,0 +1,9 @@
+Instead of defining the constant _inside_ the function, consider extracting it
```
Comment (ErikSchierboom) [inline automated-comments/javascript/gigasecond/prefer_extracted_top_level_constant.md]: Should this perhaps be a general comment, where the name of exported value is parameterized?

File: automated-comments/javascript/gigasecond/prefer_side_effect_free_date.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
+Use [`Date#getTime`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTime)
+and [`new Date(value)`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date#Syntax),
+which takes a number of milliseconds as `value` and constructs a _new_ date.
+This ensures the _input_ is not modified when calling `gigasecond`, which
+also means that there are no unintended side-effects.
+
+In general, [Unix/UTC time](https://en.wikipedia.org/wiki/Unix_time),
+as provided by `Date#getTime`, is preferred when dealing with dates, as it is
+not affected by _interpretation_ or locale (such as timezones), and it allows
+for date manupilation without needing to rely on _roll-over_ (as `setSeconds`
```
Comment (ErikSchierboom) [inline automated-comments/javascript/gigasecond/prefer_side_effect_free_date.md]: Same comments as before.

File: automated-comments/javascript/gigasecond/signature_not_optimal.md [hunk 4]
```diff
@@ -0,0 +1,6 @@
+
```
Comment (ErikSchierboom) [inline automated-comments/javascript/gigasecond/signature_not_optimal.md]:
```suggestion
```

File: automated-comments/javascript/gigasecond/signature_not_optimal.md [hunk 5]
```diff
@@ -0,0 +1,6 @@
...
+Looking at the tests, the function `gigasecond(...)` only receives one
+parameter. Remove the additional parameters from the definition of `gigasecond`,
+as the value of those extranous parameters will always be `undefined`, or always
```
Comment (ErikSchierboom) [inline automated-comments/javascript/gigasecond/signature_not_optimal.md]:
```suggestion
as the value of those extra parameters will always be `undefined`, or
```

File: automated-comments/javascript/gigasecond/prefer_extracted_top_level_constant.md (hunk 2 above)
Comment (SleeplessByte) [reply automated-comments/javascript/gigasecond/prefer_extracted_top_level_constant.md]: Perhaps, but this is sorta a special case 😅. I'll move it to general if this comes up more often.

Basically, usually I don't just say to "extract it", but "extract it because xyz". In the case of `gigasecond`, most of the time people have a literal or big number comprehension inline. In that case it uses a comment that's already in `website-copy` that explains you can extract it into a `const`. This comment is for when it's already a `const`,  but defined inside the function body. In the case of `resistor-color`, when people define it inside the function body, they re-create the colors array all the time, which is the extra commentary given there.

In other words, I think in general when we're saying to extract it to the top level, there are extra reasons why, **except** for this particular case.

File: automated-comments/javascript/gigasecond/dont_use_get_seconds.md (hunk 1 above)
Comment (SleeplessByte) [inline automated-comments/javascript/gigasecond/dont_use_get_seconds.md]: Hmmm. I guess I expect them too but they might not. How about:

```suggestion
for date manipulation without needing to rely on roll-over, which, for example, occurs when the seconds components is larger than 60, and adds one to the minute component.
```

Comment (SleeplessByte) [reply automated-comments/javascript/gigasecond/dont_use_get_seconds.md]: Hmmm. I guess I expect them too but they might not. How about:

```suggestion
for date manipulation without needing to rely on roll-over, which, for example, occurs when the seconds components is larger than 60, and adds one to the minute component.
```

Comment (ErikSchierboom) [reply automated-comments/javascript/gigasecond/dont_use_get_seconds.md]: I like the suggested change.

File: automated-comments/javascript/gigasecond/prefer_extracted_top_level_constant.md (hunk 2 above)
Comment (ErikSchierboom) [reply automated-comments/javascript/gigasecond/prefer_extracted_top_level_constant.md]: Right, that makes sense.

File: automated-comments/javascript/gigasecond/prefer_side_effect_free_date.md (hunk 3 above)
Comment (ErikSchierboom) [inline automated-comments/javascript/gigasecond/prefer_side_effect_free_date.md]:
```suggestion
for date manipulation without needing to rely on _roll-over_ (as `setSeconds`
```

Comment (SleeplessByte) [reply automated-comments/javascript/gigasecond/prefer_side_effect_free_date.md]:
```suggestion
for date manipulation without needing to rely on _roll-over_, which, for example, occurs when the seconds components is larger than 60, and adds one to the minute component (as `setSeconds`
```

File: automated-comments/javascript/gigasecond/dont_use_get_seconds.md [hunk 6]
```diff
@@ -0,0 +1,6 @@
+Use [`Date#getTime`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTime)
+to get the number of milliseconds the _entire_ `Date` represents, instead of
+getting the `seconds` component of the `Date`. In general, [Unix/UTC time](https://en.wikipedia.org/wiki/Unix_time),
+as provided by `Date#getTime`, is preferred when dealing with dates, as it is
+not affected by _interpretation_ or locale (such as timezones), and it allows
+for date manipulation without needing to rely on roll-over.
```
Comment (SleeplessByte) [inline automated-comments/javascript/gigasecond/dont_use_get_seconds.md]:
```suggestion
for date manipulation without needing to rely on roll-over, which, for example, occurs when the seconds components is larger than 60, and adds one to the minute component.
```

**Comment by SleeplessByte** [comment]
Thank you both!
